cmd/gossamer/config.go

```diff
@@ -287,115 +288,98 @@ func createExportConfig(ctx *cli.Context) (*dot.Config, error) {
 	return cfg, nil
 }
 
-func setLogConfig(ctx *cli.Context, cfg *ctoml.Config, globalCfg *dot.GlobalConfig, logCfg *dot.LogConfig) error {
-	if cfg == nil {
-		cfg = new(ctoml.Config)
-	}
-
-	if lvlStr := ctx.String(LogFlag.Name); lvlStr != "" {
-		if lvlToInt, err := strconv.Atoi(lvlStr); err == nil {
-			lvlStr = log.Lvl(lvlToInt).String()
-		}
-		cfg.Global.LogLvl = lvlStr
-	}
+type getStringer interface {
+	String(key string) (value string)
+}
 
-	if cfg.Global.LogLvl == "" {
-		cfg.Global.LogLvl = gssmr.DefaultLvl.String()
+// getLogLevel obtains the log level in the following order:
+// 1. Try to obtain it from the flag value corresponding to flagName.
+// 2. Try to obtain it from the TOML value given, if step 1. failed.
+// 3. Return the default value given if both previous steps failed.
+// For steps 1 and 2, it tries to parse the level as an integer to convert it
+// to a level, and also tries to parse it as a string.
+func getLogLevel(ctx getStringer, flagName, tomlValue string, defaultLevel log.Lvl) (
+	level log.Lvl, err error) {
+	if flagValue := ctx.String(flagName); flagValue != "" {
+		return parseLogLevelString(flagValue)
 	}
 
-	var err error
-	globalCfg.LogLvl, err = log.LvlFromString(cfg.Global.LogLvl)
-	if err != nil {
-		return err
+	if tomlValue == "" {
+		return defaultLevel, nil
 	}
 
-	// check and set log levels for each pkg
-	if cfg.Log.CoreLvl == "" {
-		logCfg.CoreLvl = globalCfg.LogLvl
-	} else {
-		lvl, err := log.LvlFromString(cfg.Log.CoreLvl)
-		if err != nil {
-			return err
-		}
+	return parseLogLevelString(tomlValue)
+}
 
-		logCfg.CoreLvl = lvl
-	}
+var regexDigits = regexp.MustCompile("^[0-9]+$")
 
-	if cfg.Log.SyncLvl == "" {
-		logCfg.SyncLvl = globalCfg.LogLvl
-	} else {
-		lvl, err := log.LvlFromString(cfg.Log.SyncLvl)
-		if err != nil {
-			return err
+func parseLogLevelString(logLevelString string) (logLevel log.Lvl, err error) {
+	if regexDigits.MatchString(logLevelString) {
+		levelInt, err := strconv.Atoi(logLevelString)
+		if err != nil { // should never happen
+			return 0, fmt.Errorf("cannot parse log level digits: %w", err)
 		}
+		logLevel = log.Lvl(levelInt)
+		return logLevel, nil
+	}
```

Comment:
is the regex needed? can you just do `Atoi` and if the int is <0 or >4 then error? although, I guess the following lines achieve the same thing?

Reply:
I put this up there mostly for readability. Reverted with extra comment.

I also added a check for the integer value to be between `0` and `5` with unit test.